src/shared/install.c

```diff
@@ -66,6 +69,31 @@ typedef struct {
         OrderedHashmap *have_processed;
 } InstallContext;
 
+typedef struct {
+        const char *pattern;
+        int action;
+} PresetRule;
+
+typedef struct {
+        PresetRule *rules;
+        size_t n_rules;
+} Presets;
+
+static inline void presets_freep(Presets *p) {
+        size_t i;
+
+        if (p) {
+                for (i = 0; i < p->n_rules; i++)
+                        if (p->rules[i].pattern)
+                                free((char *)p->rules[i].pattern);
+
+                free(p->rules);
+                p->n_rules = 0;
+        }
+}
+
+#define _cleanup_presets_ _cleanup_(presets_freep)
+
```

Comment:
I am not 100% sure but it seems to me this cleanup handling is inconsistent with rest of the code base. I think it would be better to heap allocate ```Presets``` and implement cleanup handler in a usual way, i.e. rename current ```preset_freep``` to ```presets_free``` and define trivial cleanup handler ```presents_freep``` using standard ```DEFINE_TRIVIAL_CELANUP_FUNC```.